### PR TITLE
Kernel: Ignore interfaces without an IP address when updating ARP

### LINKED ISSUE #13713: Kernel: ARP table caches public IP addresses outside of LAN
ARP is a layer two protocol really meant for devices to learn about each other on the same local network. The system currently adds an ARP table entry for _any_ IP address (private or public) it interacts with, while it should instead only cache the Kernel's IPv4 network neighbors.

It's more apparent when you browse to multiple public web sites, and anything outside of the LAN (the public IP addresses) are all entered with the same local gateway MAC address. 

serenity_arp

### PR DESCRIPTION
**Pull request by brapru**
For the same reason we ignore interfaces without an IP address when
choosing where to send a route, we should also ignore interfaces without
IP addresses when updating the ARP table on incoming packets from
local addresses.

On an interface with a null address, the mask checking would always
result in zero, which resulted in the system updating the ARP table on
almost every incoming packet from any address (private or public).

This patch fixes this behavior by only applying this check to interfaces
with valid addresses and now the ARP table won't get constantly
hammered.

Closes #13713

### CODE REVIEW

File: Kernel/Net/NetworkTask.cpp [hunk 1]
```diff
@@ -202,7 +202,7 @@ void handle_ipv4(EthernetFrameHeader const& eth, size_t frame_size, Time const&
     dbgln_if(IPV4_DEBUG, "handle_ipv4: source={}, destination={}", packet.source(), packet.destination());
 
     NetworkingManagement::the().for_each([&](auto& adapter) {
-        if (adapter.link_up()) {
+        if (adapter.link_up() && (adapter.ipv4_address().to_u32() != 0)) {
```
Comment (IdanHo) [inline Kernel/Net/NetworkTask.cpp:205]: nitpick, but can we flip the condition to reduce the level of indentation a bit?

Comment (brapru) [reply Kernel/Net/NetworkTask.cpp:205]: Sure thing. Actually, we can leverage AK/IPv4Address here too. How about this?

`if (adapter.link_up() && !adapter.ipv4_address().is_zero())`

Comment (IdanHo) [reply Kernel/Net/NetworkTask.cpp:205]: Sure, with the condition flipped though :^)

Comment (brapru) [reply Kernel/Net/NetworkTask.cpp:205]: Haha I'm having a slow moment. Are you saying like this?

`if (!adapter.ipv4_address().is_zero() && adapter.link_up())`

Or do you mean just put it in a separate conditional statement in the beginning and just return immediately:

```
 if (adapter.ipv4_address().is_zero())
    return
```

Comment (IdanHo) [reply Kernel/Net/NetworkTask.cpp:205]: The second option, but include `|| !adapter.link_up()` as well, the `return early` pattern usually results in flatter and easier-to-understand code IMO

Comment (brapru) [reply Kernel/Net/NetworkTask.cpp]: Let me know if that's better. Thanks!

**Comment by IdanHo** [comment]
Hmm, `TestTLSHandshake` failed in CI, that seems vaguely related, not sure if it's caused by this

**Comment by brapru** [comment]
> Hmm, `TestTLSHandshake` failed in CI, that seems vaguely related, not sure if it's caused by this

Yeah this one seems to be struggling with working through CI, I think it's more of the flakiness. I'm able to pass all tests on clean builds. Also it was failing at different parts of the pipeline on previous pushes. 

In any case, here's another shot with a rebase ¯\_(ツ)_/¯

**Comment by brapru** [comment]
> Make sure youre running the exact configuration that specific CI run is using, e.g. `i686` + KVM disabled + CI run mode (no graphics, etc)

Hmm. Yeah, I've stepped through the CI pipeline exactly on clean builds. Still having trouble replicating the timeout :confused:

**Comment by brapru** [comment]
I’ve been able to peel back a couple of layers. I will say I thought this was a simple change and I've gone down a much longer rabbit hole than intended.

A couple of things to note:
- I cannot reproduce it on my local system. I’ve created the build step by step mirroring the CI steps and I pass the test every time. 
- I can reproduce the behavior when I build inside a docker container. I figured this would more resemble how it would run inside Github actions, and the LibTLS/TestTLSHandshake indeed does hang. 
- If I run the TestTLSHandshake individually, it will pass every time. It only hangs when you run the whole run-tests-and-shutdown.sh script.
- Through some print-fu debugging, the `TLSv12::connect` -> `Core::Stream::TCPSocket::connect` never returns when it calls `resolve_host`. It seems to hang around [here](https://github.com/SerenityOS/serenity/blob/e815d3f9ce0a0e0371c0852d9b911fe0611d9f3b/Userland/Libraries/LibCore/Stream.cpp#L303) and doesn't return.

All this being said, I’m struggling to see the relationship between the commit changes and this behavior. @IdanHo, any thoughts or recommendations? Thanks!

**Pull request merged**